s3/aide.go

```diff
@@ -1,87 +1,95 @@
 package s3
 
 import (
-	"encoding/json"
-	"io"
-	"io/ioutil"
+	"bytes"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/aws/aws-sdk-go/service/s3/s3iface"
-	"github.com/cleardataeng/aidews"
 )
 
-// Service for reading and writing to the given bucket.
-type Service struct {
-	// acl is the default ACL for bucket objects.
-	acl *string
+// Bucket allows access and control of an S3 bucket.
+type Bucket struct {
+	// Name of the bucket.
+	Name string
 
-	// name of the bucket.
-	name string
+	// Svc is the S3API client.
+	// If not provided, default credential resolution will take place.
+	Svc s3iface.S3API
+}
 
-	// sse is the default server side encryption setting.
-	sse *string
+// Object allows access to S3 objects.
+type Object struct {
+	// ACL is the ACL for an object.
+	// If empty "bucket-owner-full-control" will be used.
+	ACL string
 
-	// svc is the S3API client.
-	svc s3iface.S3API
-}
+	// Bucket in which the object belongs.
+	Bucket Bucket
 
-// New returns a pointer to a new Service.
-// ACL is bucket-owner-full-control by default, but can be changed with SetACL.
-// SSE is AES256 by default, but can be changed with SetSSE.
-func New(name string, region, roleARN *string) *Service {
-	return &Service{
-		acl:  aws.String("bucket-owner-full-control"),
-		name: name,
-		sse:  aws.String("AES256"),
-		svc:  s3.New(aidews.Session(region, roleARN)),
-	}
+	// Key at which the object is or shall be stored.
+	Key string
+
+	// SSE is the server side encryption setting.
+	// If empty "AES256" will be used.
+	SSE string
 }
 
-// Put puts the content to the bucket at the key.
-func (svc *Service) Put(key string, content io.Reader) (*s3.PutObjectOutput, error) {
-	in := &s3.PutObjectInput{
-		ACL:                  svc.acl,
-		Body:                 aws.ReadSeekCloser(content),
-		Bucket:               aws.String(svc.name),
-		Key:                  aws.String(key),
-		ServerSideEncryption: svc.sse,
+// Objects is a simple constructor helper function taking a list of
+// strings, S3 keys, and generating []Object.
+// The idea is to get a list of objects from the main AWS SDK, using
+// the full blow input, then passing the returned keys to this function.
+// From there, since each object is implementing io.Reader, you can
```

Comment:
This comment may want to reflect that since the objects implement writer that a bunch of keys can get written to as well this way.